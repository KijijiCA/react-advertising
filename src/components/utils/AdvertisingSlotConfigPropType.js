import PropTypes from 'prop-types';

export default PropTypes.shape({
    id: PropTypes.string.isRequired,
    path: PropTypes.string,
    collapseEmptyDiv: PropTypes.arrayOf(PropTypes.bool),
    targeting: PropTypes.object,
    sizes: PropTypes.oneOfType([
        PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.number)]),
        PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.number)])),
    ]),
    sizeMappingName: PropTypes.string,
    prebid: PropTypes.arrayOf(
        PropTypes.shape({
            mediaTypes: PropTypes.objectOf(
                PropTypes.shape({
                    sizes: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)),
                })
            ).isRequired,
            bids: PropTypes.arrayOf(
                PropTypes.shape({
                    bidder: PropTypes.string.isRequired,
                    params: PropTypes.object,
                    labelAny: PropTypes.arrayOf(PropTypes.string),
                    labelAll: PropTypes.arrayOf(PropTypes.string),
                })
            ).isRequired,
        })
    ).isRequired,
});
