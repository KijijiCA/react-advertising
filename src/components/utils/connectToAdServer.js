import React from 'react';
import AdvertisingContext from '../../AdvertisingContext';

export default (Component) => (props) => (
    <AdvertisingContext.Consumer>
        {(activate) => <Component {...props} activate={activate} />}
    </AdvertisingContext.Consumer>
);
