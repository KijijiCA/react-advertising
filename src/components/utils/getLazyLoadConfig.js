export default function getLazyLoadConfig(config, id) {
  if (!config?.slots) {
    return null;
  }
  const slotConfig = config.slots.find((slot) => slot.id === id);
  if (!slotConfig) {
    return null;
  }
  if (
    slotConfig.enableLazyLoad !== undefined &&
    slotConfig.enableLazyLoad !== null
  ) {
    return slotConfig.enableLazyLoad;
  }
  return config?.enableLazyLoad === undefined ? null : config?.enableLazyLoad;
}
