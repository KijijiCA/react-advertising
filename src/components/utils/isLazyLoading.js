export default function isLazyLoading(lazyLoadConfig) {
  if (typeof lazyLoadConfig === 'boolean') {
    return lazyLoadConfig;
  }

  if (!lazyLoadConfig) {
    return false;
  }

  const { marginPercent } = lazyLoadConfig;

  if (!marginPercent) {
    return true;
  }

  return marginPercent > -1;
}
