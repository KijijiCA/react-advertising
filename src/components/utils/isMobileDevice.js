// Mobile device detection, based on MDN solution:
// https://developer.mozilla.org/en-US/docs/Web/HTTP/Browser_detection_using_the_user_agent
export default function isMobileDevice() {
  if ('maxTouchPoints' in navigator) {
    return navigator.maxTouchPoints > 0;
  }
  if ('msMaxTouchPoints' in navigator) {
    return navigator.msMaxTouchPoints > 0;
  }
  const mQ = window.matchMedia && matchMedia('(pointer:coarse)');
  if (mQ && mQ.media === '(pointer:coarse)') {
    return !!mQ.matches;
  }
  if ('orientation' in window) {
    return true; // deprecated, but good fallback
  }
  // Only as a last resort, fall back to user agent sniffing
  const UA = navigator.userAgent;
  return (
    /\b(BlackBerry|webOS|iPhone|IEMobile)\b/i.test(UA) ||
    /\b(Android|Windows Phone|iPad|iPod)\b/i.test(UA)
  );
}
