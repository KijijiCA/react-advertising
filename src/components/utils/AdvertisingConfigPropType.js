import PropTypes from 'prop-types';
import AdvertisingSlotConfigPropType from './AdvertisingSlotConfigPropType';

export default PropTypes.shape({
  path: PropTypes.string,
  targeting: PropTypes.object,
  usePrebid: PropTypes.bool,
  prebid: PropTypes.shape({
    debug: PropTypes.bool,
    bidderTimeout: PropTypes.number,
    enableSendAllBids: PropTypes.bool,
    bidderSequence: PropTypes.oneOf(['random', 'fixed']),
    publisherDomain: PropTypes.string,
    cookieSyncDelay: PropTypes.number,
    priceGranularity: PropTypes.oneOfType([
      PropTypes.oneOf(['low', 'medium', 'high', 'auto', 'dense']),
      PropTypes.shape({
        buckets: PropTypes.arrayOf(
          PropTypes.shape({
            precision: PropTypes.number,
            min: PropTypes.number.isRequired,
            max: PropTypes.number.isRequired,
            increment: PropTypes.number.isRequired,
          })
        ).isRequired,
      }),
    ]),
    mediaTypePriceGranularity: PropTypes.shape({
      video: PropTypes.oneOfType([
        PropTypes.oneOf(['low', 'medium', 'high', 'auto', 'dense']),
        PropTypes.shape({
          buckets: PropTypes.arrayOf(
            PropTypes.shape({
              precision: PropTypes.number,
              min: PropTypes.number.isRequired,
              max: PropTypes.number.isRequired,
              increment: PropTypes.number.isRequired,
            })
          ).isRequired,
        }),
      ]),
      banner: PropTypes.oneOfType([
        PropTypes.oneOf(['low', 'medium', 'high', 'auto', 'dense']),
        PropTypes.shape({
          buckets: PropTypes.arrayOf(
            PropTypes.shape({
              precision: PropTypes.number,
              min: PropTypes.number.isRequired,
              max: PropTypes.number.isRequired,
              increment: PropTypes.number.isRequired,
            })
          ).isRequired,
        }),
      ]),
      native: PropTypes.oneOfType([
        PropTypes.oneOf(['low', 'medium', 'high', 'auto', 'dense']),
        PropTypes.shape({
          buckets: PropTypes.arrayOf(
            PropTypes.shape({
              precision: PropTypes.number,
              min: PropTypes.number.isRequired,
              max: PropTypes.number.isRequired,
              increment: PropTypes.number.isRequired,
            })
          ).isRequired,
        }),
      ]),
    }),
    sizeConfig: PropTypes.arrayOf(
      PropTypes.shape({
        mediaQuery: PropTypes.string.isRequired,
        sizesSupported: PropTypes.arrayOf(
          PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.arrayOf(PropTypes.number),
          ])
        ),
        labels: PropTypes.arrayOf(PropTypes.string),
      })
    ),
  }),
  sizeMappings: PropTypes.objectOf(
    PropTypes.arrayOf(
      PropTypes.shape({
        viewPortSize: PropTypes.arrayOf(PropTypes.number).isRequired,
        sizes: PropTypes.arrayOf(
          PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.arrayOf(PropTypes.number),
          ])
        ).isRequired,
      })
    )
  ),
  slots: PropTypes.arrayOf(AdvertisingSlotConfigPropType),
  outOfPageSlots: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
    })
  ),
  customEvents: PropTypes.objectOf(
    PropTypes.shape({
      eventMessagePrefix: PropTypes.string.isRequired,
      divIdPrefix: PropTypes.string,
    })
  ),
  enableLazyLoad: PropTypes.oneOfType([
    PropTypes.bool,
    PropTypes.shape({
      marginPercent: PropTypes.number,
      mobileScaling: PropTypes.number,
    }),
  ]),
});
