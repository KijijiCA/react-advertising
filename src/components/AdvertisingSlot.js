import React, { Component } from 'react';
import connectToAdServer from './utils/connectToAdServer';
import PropTypes from 'prop-types';

class AdvertisingSlot extends Component {
    componentDidMount() {
        const { activate, id, customEventHandlers } = this.props;
        activate(id, customEventHandlers);
    }

    componentDidUpdate(prevProps) {
        const { activate, id, customEventHandlers } = this.props;
        if (prevProps.activate !== activate) {
            activate(id, customEventHandlers);
        }
    }

    render() {
        const { id, style, className, children } = this.props;
        return <div id={id} style={style} className={className} children={children} />;
    }
}

AdvertisingSlot.propTypes = {
    id: PropTypes.string.isRequired,
    style: PropTypes.object,
    className: PropTypes.string,
    children: PropTypes.node,
    activate: PropTypes.func.isRequired,
    customEventHandlers: PropTypes.objectOf(PropTypes.func).isRequired
};

AdvertisingSlot.defaultProps = {
    customEventHandlers: {}
};

export default connectToAdServer(AdvertisingSlot);
