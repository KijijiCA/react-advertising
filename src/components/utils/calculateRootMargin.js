import isMobileDevice from './isMobileDevice';

export default function calculateRootMargin({
  marginPercent,
  mobileScaling,
} = {}) {
  if (!marginPercent) {
    return undefined;
  }
  const finalMarginPercent =
    isMobileDevice() && mobileScaling !== undefined && mobileScaling !== -1
      ? marginPercent * mobileScaling
      : marginPercent;
  return `${(finalMarginPercent / 100) * window.innerHeight}px`;
}
