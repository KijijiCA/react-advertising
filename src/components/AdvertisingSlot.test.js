import React from 'react';
import { render } from '@testing-library/react';
import AdvertisingSlot from './AdvertisingSlot';
import AdvertisingContext from '../AdvertisingContext';
import { config, DIV_ID_FOO } from '../utils/testAdvertisingConfig';

const mockActivate = jest.fn();

describe('The advertising slot component', () => {
  let slot, rerender;

  beforeEach(() => {
    jest.clearAllMocks();
    ({
      container: { firstChild: slot },
      rerender,
    } = render(
      <AdvertisingContext.Provider
        value={{
          activate: mockActivate,
          config,
        }}
      >
        <AdvertisingSlot
          id={DIV_ID_FOO}
          style={{ color: 'hotpink' }}
          className="my-class"
        >
          <h1>hello</h1>
        </AdvertisingSlot>
      </AdvertisingContext.Provider>
    ));
  });

  it('renders correctly', () => {
    expect(slot).toMatchSnapshot();
  });

  it('calls the activate function with the ID', () => {
    expect(mockActivate).toHaveBeenCalledWith(DIV_ID_FOO, expect.anything());
  });

  it('calls the activate function with a collapse callback', () => {
    expect(mockActivate).toHaveBeenCalledWith(
      expect.anything(),
      expect.any(Object)
    );
  });

  it('calls the new activate function if the new activate function changes', (done) => {
    const newMockActivate = jest.fn();
    rerender(
      <AdvertisingContext.Provider
        value={{
          activate: newMockActivate,
          config,
        }}
      >
        <AdvertisingSlot id={DIV_ID_FOO} />
      </AdvertisingContext.Provider>
    );

    setTimeout(() => {
      expect(newMockActivate).toHaveBeenCalledTimes(1);
      done();
    }, 0);
  });

  it('does not call the new activate function if the activate function does not change', (done) => {
    rerender(
      <AdvertisingContext.Provider
        value={{
          activate: mockActivate,
          config,
        }}
      >
        <AdvertisingSlot id={DIV_ID_FOO} />
      </AdvertisingContext.Provider>
    );

    setTimeout(() => {
      expect(mockActivate).toHaveBeenCalledTimes(1);
      done();
    }, 0);
  });
});
