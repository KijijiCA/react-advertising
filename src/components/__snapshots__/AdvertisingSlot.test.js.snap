// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The advertising slot component renders correctly 1`] = `
<div
  class="my-class"
  id="div-gpt-ad-foo"
  style="color: hotpink;"
>
  <h1>
    hello
  </h1>
</div>
`;
